staging/src/k8s.io/apiserver/pkg/storage/interfaces.go

```diff
@@ -264,6 +264,18 @@ type Interface interface {
 	RequestWatchProgress(ctx context.Context) error
 }
 
+// Decoder is used by the storage implementation to decode transformed
+// data from the storage into an object
+type Decoder interface {
+	// Decode decodes value of bytes into object. It will also
+	// set the object resource version to rev.
+	// On success, objPtr would be set to the object.
+	Decode(codec runtime.Codec, versioner Versioner, value []byte, objPtr runtime.Object, rev int64) error
```

Comment:
how generic is that really with a int64 rv? do we have this convention elsewhere yet on the storage level?

Reply:
Alternative: make rev a string, and let the etcd3 store do the conversion.

Reply:
and then this is really the resource version, not an etcd rev.

Reply:
and then we are not far from moving the while default decoder here 🤔

Reply:
it's not generic, it applies to etcd implementation only, I have put the interface definition in the storage folder for convenience, maybe we can put the interface definition in the `storage/etcd3` folder, let me look into it.

Reply:
okay now it's `etcd3.Decoder`, not `storage.Decoder`